File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function insertGreatestCommonDivisors(head: ListNode | null): ListNode | null {
  const gcd = (a: number, b :number):number => {
    while (a > 0 && b > 0) [a, b] = [Math.min(a, b), Math.max(a, b) % Math.min(a, b)];
    return Math.max(a, b);
  }
  let node = head;
  while (node) {
    if (!node.next) break;
    node.next = new ListNode(gcd(node.val, node.next.val), node.next);
    node = node.next.next;
  }
  return head;
};

File: 2807-insert-greatest-common-divisors-in-linked-list/2807-insert-greatest-common-divisors-in-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertGreatestCommonDivisors(head *ListNode) *ListNode {
  node := head
  for node != nil {
    if node.Next == nil {
      break
    }
    a, b := node.Val, node.Next.Val
    if a != b {
      for a != 0 && b != 0 {
        a, b = a % b, b % a
      }
    }
    node.Next = &ListNode{ Val: max(a, b), Next: node.Next }
    node = node.Next.Next
  }
  return head
}